File: solutions/graphs/alternatives/997-find-the-town-judge.js
/**

 *
 * This problem demonstrates key concepts in Graphs.
 *
 * SOLUTION EXPLANATION:
 *
 * INTUITION:
 * The town judge is trusted by everyone but trusts no one.
 * In graph terms: judge has indegree n-1 and outdegree 0.
 *
 * APPROACH:



 *
 * WHY THIS WORKS:
 * - Judge: receives n-1 trust, gives 0 trust → net = n-1
 * - Others: give at least 1 trust to judge → net ≤ n-2
 * - Only one person can have net score of n-1
 *
 * TIME COMPLEXITY: O(E) where E is number of trust relationships
 * SPACE COMPLEXITY: O(n) for trust score array
 *
 * EXAMPLE WALKTHROUGH:
 * n=3, trust=[[1,3],[2,3]]
 * Person 1: gives 1, receives 0 → net = -1
 * Person 2: gives 1, receives 0 → net = -1
 * Person 3: gives 0, receives 2 → net = 2 = n-1
 * Judge: 3
 *
 * EDGE CASES:
 * - Single person (automatically judge)
 * - No judge exists (multiple candidates or none)
 * - Empty trust relationships
 * - Everyone trusts everyone
 */

/**
 * Main solution for Problem 997: Find The Town Judge
 *
 * @param {number} n - Number of people labeled 1 to n
 * @param {number[][]} trust - Array of trust relationships [a, b] means a trusts b
 * @return {number} - The town judge's label, or -1 if no judge exists
 *
 * Time Complexity: O(E) where E is number of trust relationships
 * Space Complexity: O(n) for trust score array
 */
function solve(n, trust) {
    // Special case: single person is automatically the judge
    if (n === 1) return 1;

    // Calculate net trust score for each person (indegree - outdegree)
    const trustScore = new Array(n + 1).fill(0);

    for (const [a, b] of trust) {
        trustScore[a]--; // a trusts someone (outdegree)
        trustScore[b]++; // b is trusted by someone (indegree)
    }

    // Find person with trust score of n-1 (trusted by all others, trusts no one)
    for (let i = 1; i <= n; i++) {
        if (trustScore[i] === n - 1) {
            return i;
        }
    }

    return -1; // No judge found
}

/**
 * Test cases for Problem 997: Find The Town Judge
 */
function testSolution() {
    console.log('Testing 997. Find The Town Judge');

    // Test case 1: Basic case with judge
    const result1 = solve(2, [[1,2]]);
    console.assert(result1 === 2, `Test 1 failed: expected 2, got ${result1}`);

    // Test case 2: Three people, person 3 is judge
    const result2 = solve(3, [[1,3],[2,3]]);
    console.assert(result2 === 3, `Test 2 failed: expected 3, got ${result2}`);

    // Test case 3: No judge (person 3 trusts someone)
    const result3 = solve(3, [[1,3],[2,3],[3,1]]);
    console.assert(result3 === -1, `Test 3 failed: expected -1, got ${result3}`);

    // Test case 4: Single person
    const result4 = solve(1, []);
    console.assert(result4 === 1, `Test 4 failed: expected 1, got ${result4}`);

    // Test case 5: No trust relationships
    const result5 = solve(3, []);
    console.assert(result5 === -1, `Test 5 failed: expected -1, got ${result5}`);

    // Test case 6: Multiple people trust each other (no judge)
    const result6 = solve(3, [[1,2],[2,3],[3,1]]);
    console.assert(result6 === -1, `Test 6 failed: expected -1, got ${result6}`);

    // Test case 7: Judge doesn't trust everyone else
    const result7 = solve(4, [[1,2],[1,3],[2,3],[3,4],[4,3]]);
    console.assert(result7 === -1, `Test 7 failed: expected -1, got ${result7}`);

    // Test case 8: Large case with clear judge
    const result8 = solve(4, [[1,4],[2,4],[3,4]]);
    console.assert(result8 === 4, `Test 8 failed: expected 4, got ${result8}`);

    console.log('All test cases passed for 997. Find The Town Judge!');
}

/**
 * Example usage and demonstration
 */
function demonstrateSolution() {
    console.log('\n=== Problem 997. Find The Town Judge ===');
    console.log('Category: Graphs');
    console.log('Difficulty: Easy');
    console.log('');

    // Example demonstration would go here
    testSolution();
}

// Run tests if this file is executed directly
if (require.main === module) {
    demonstrateSolution();
}

// Export for use in other modules
module.exports = {
    solve,
    testSolution,
    demonstrateSolution
};

/**
 * Additional Notes:
 * - This solution focuses on graphs concepts
 * - Consider the trade-offs between time and space complexity
 * - Edge cases are crucial for robust solutions
 * - The approach can be adapted for similar problems in this category
 */

File: solutions/graphs/997-find-the-town-judge.py
"""
# Difficulty: Easy

# 997. Find The Town Judge

In a town, there are n people labeled from 1 to n. There is a rumor that one of these people is secretly the town judge.

If the town judge exists, then:




You are given an array trust where trust[i] = [ai, bi] representing that the person labeled ai trusts the person labeled bi.

Return the label of the town judge if the town judge exists and can be identified, or return -1 otherwise.

**Example:**

<dl class="example-details">
<dt>Input:</dt>
<dd>[(2, [[1,2]</dd>
<dt>Output:</dt>
<dd>"n={n}, trust={trust} -> Judge: {result}"</dd>
<dt>Explanation:</dt>
<dd>The town judge is person 2 who is trusted by all but trusts no one</dd>
</dl>

<details>
<summary><b>🔍 SOLUTION EXPLANATION</b></summary>


### METADATA:
**Techniques**: TBD
**Data Structures**: TBD
**Patterns**: TBD
**Time Complexity**: O(T + N)
**Space Complexity**: O(N)

### INTUITION:
This is a graph problem where we need to find a node (person) with specific in-degree and out-degree properties. The judge must have in-degree = n-1 (everyone trusts them) and out-degree = 0 (they trust nobody). We can solve this efficiently by tracking trust relationships as a directed graph.

### APPROACH:
1. **Count trust relationships**: Track who trusts whom and who is trusted by whom
2. **Calculate net trust**: For each person, calculate (trusted_by_count - trusts_count)
3. **Find the judge**: The judge will have net trust = n-1 (trusted by n-1 people, trusts 0)
4. **Validate result**: Ensure exactly one person satisfies the judge criteria

### WHY THIS WORKS:
- Judge trusts nobody: out-degree = 0
- Everyone else trusts judge: in-degree = n-1
- Net trust = in-degree - out-degree = (n-1) - 0 = n-1
- All other people have net trust < n-1 (they either trust someone or aren't trusted by everyone)

### EXAMPLE WALKTHROUGH:
```
Input: n = 3, trust = [[1,3],[2,3]]
Person 1: trusts 3, trusted by 0 → net = 0 - 1 = -1
Person 2: trusts 3, trusted by 0 → net = 0 - 1 = -1
Person 3: trusts 0, trusted by 2 → net = 2 - 0 = 2 = n-1 ✓
Output: 3 (person 3 is the judge)
```

### TIME COMPLEXITY:
O(T + N)
Where T is the number of trust relationships and N is the number of people

### SPACE COMPLEXITY:
O(N)
For storing trust counts

### EDGE CASES:
- n = 1: Only one person, they are the judge by default
- Empty trust array with n > 1: No judge possible
- Multiple people with high trust: No unique judge
- Circular trust: No judge possible

</details>
"""


from typing import Any
import re


class Solution:
    def findJudge(self, n: int, trust: list[list[int]]) -> int:
        """
        Find the town judge using net trust calculation.

        Args:
            n: Number of people (labeled 1 to n)
            trust: List of trust relationships [a, b] meaning a trusts b

        Returns:
            Label of the town judge, or -1 if no judge exists

        Time Complexity: O(T + N) where T is trust relationships, N is people
        Space Complexity: O(N) for trust count arrays
        """
        # Special case: only one person
        if n == 1:
            return 1 if not trust else -1

        # Track net trust: (trusted_by_count - trusts_count) for each person
        net_trust = [0] * (n + 1)  # Index 0 unused, 1 to n for people

        # Process all trust relationships
        for truster, trustee in trust:
            net_trust[truster] -= 1  # Truster loses trust (trusts someone)
            net_trust[trustee] += 1  # Trustee gains trust (is trusted by someone)

        # Find person with net trust = n-1 (judge criteria)
        for person in range(1, n + 1):
            if net_trust[person] == n - 1:
                return person

        return -1  # No judge found

    def findJudgeAlternative(self, n: int, trust: list[list[int]]) -> int:
        """
        Alternative implementation using separate in-degree and out-degree tracking.

        Args:
            n: Number of people
            trust: Trust relationships

        Returns:
            Judge label or -1
        """
        if n == 1:
            return 1

        # Track in-degree (trusted by) and out-degree (trusts)
        trusted_by = [0] * (n + 1)  # How many people trust this person
        trusts = [0] * (n + 1)  # How many people this person trusts

        for truster, trustee in trust:
            trusts[truster] += 1
            trusted_by[trustee] += 1

        # Judge must be trusted by n-1 people and trust nobody
        for person in range(1, n + 1):
            if trusted_by[person] == n - 1 and trusts[person] == 0:
                return person

        return -1

    def findJudgeOptimized(self, n: int, trust: list[list[int]]) -> int:
        """
        Space-optimized solution using single pass and candidate tracking.

        Args:
            n: Number of people
            trust: Trust relationships

        Returns:
            Judge label or -1
        """
        if n == 1:
            return 1 if not trust else -1

        # Use a different approach: find candidate with highest net trust
        trust_score: dict[Any, Any] = {}

        for truster, trustee in trust:
            trust_score[truster] = trust_score.get(truster, 0) - 1
            trust_score[trustee] = trust_score.get(trustee, 0) + 1

        # Check if any person has the judge criteria
        for person in range(1, n + 1):
            if trust_score.get(person, 0) == n - 1:
                return person

        return -1


def test_solution() -> None:
    """Test cases for Problem 997."""
    solution = Solution()

    # Test case 1: Basic judge scenario
    result1 = solution.findJudge(2, [[1, 2]])
    expected1 = 2
    assert result1 == expected1, f"Expected {expected1}, got {result1}"

    # Test case 2: Three people, clear judge
    result2 = solution.findJudge(3, [[1, 3], [2, 3]])
    expected2 = 3
    assert result2 == expected2, f"Expected {expected2}, got {result2}"

    # Test case 3: No judge (circular trust)
    result3 = solution.findJudge(3, [[1, 3], [2, 3], [3, 1]])
    expected3 = -1
    assert result3 == expected3, f"Expected {expected3}, got {result3}"

    # Test case 4: Single person (is judge by default)
    result4 = solution.findJudge(1, [])
    expected4 = 1
    assert result4 == expected4, f"Expected {expected4}, got {result4}"

    # Test case 5: Single person but has trust relationship (impossible)
    result5 = solution.findJudge(1, [[1, 1]])
    expected5 = -1
    assert result5 == expected5, f"Expected {expected5}, got {result5}"

    # Test case 6: No trust relationships with multiple people
    result6 = solution.findJudge(3, [])
    expected6 = -1
    assert result6 == expected6, f"Expected {expected6}, got {result6}"

    # Test case 7: Complex scenario - multiple trust but no clear judge
    result7 = solution.findJudge(4, [[1, 3], [1, 4], [2, 3], [2, 4], [4, 3]])
    expected7 = 3
    assert result7 == expected7, f"Expected {expected7}, got {result7}"

    # Test case 8: Everyone trusts each other (no judge)
    result8 = solution.findJudge(3, [[1, 2], [2, 3], [3, 1]])
    expected8 = -1
    assert result8 == expected8, f"Expected {expected8}, got {result8}"

    # Test alternative implementation
    result9 = solution.findJudgeAlternative(3, [[1, 3], [2, 3]])
    expected9 = 3
    assert result9 == expected9, f"Expected {expected9}, got {result9}"

    # Test optimized implementation
    result10 = solution.findJudgeOptimized(3, [[1, 3], [2, 3]])
    expected10 = 3
    assert result10 == expected10, f"Expected {expected10}, got {result10}"

    print("All test cases passed!")


if __name__ == "__main__":
    test_solution()

    # Example usage
    solution = Solution()
    print("=== 997. Find The Town Judge ===")

    # Demonstrate various scenarios
    test_cases = [
        (2, [[1, 2]]),
        (3, [[1, 3], [2, 3]]),
        (3, [[1, 3], [2, 3], [3, 1]]),
        (1, []),
        (4, [[1, 3], [1, 4], [2, 3], [2, 4], [4, 3]]),
    ]

    for n, trust in test_cases:
        result = solution.findJudge(n, trust)
        print(f"n={n}, trust={trust} -> Judge: {result}")

    print("\nDetailed example:")
    print("n=3, trust=[[1,3],[2,3]]")
    print("Person 1: trusts person 3 (net trust = -1)")
    print("Person 2: trusts person 3 (net trust = -1)")
    print("Person 3: trusted by persons 1,2 (net trust = +2 = n-1)")
    print("Judge: Person 3 ✓")
